Simplified/NYPLNetworkQueue.swift

```diff
@@ -0,0 +1,173 @@
+  private static let sqlHeader = Expression<Data?>("requestHeader")
+  private static let sqlRetries = Expression<Int>("retryCount")
+  
+  
+  // MARK: - Public Functions
+  
+  class func addRequest(_ libraryID: Int,
+                        _ updateID: String?,
+                        _ requestUrl: URL,
+                        _ method: HTTPMethodType,
+                        _ parameters: Data?,
+                        _ headers: [String : String]?) -> Void
+  {
+    // Serialize Data
+    let urlString = requestUrl.absoluteString
+    let methodString = method.rawValue
+    
+    let headerData: Data?
+    if headers != nil {
+      headerData = NSKeyedArchiver.archivedData(withRootObject: headers!)
+    } else {
+      headerData = nil
+    }
+    
+    guard let db = startDatabaseConnection() else { return }
+    
+    // Get or create table
+    do {
+      try db.run(sqlTable.create(ifNotExists: true) { t in
+        t.column(sqlID, primaryKey: true)
+        t.column(sqlLibraryID)
+        t.column(sqlUpdateID)
+        t.column(sqlUrl)
+        t.column(sqlMethod)
+        t.column(sqlParameters)
+        t.column(sqlHeader)
+        t.column(sqlRetries)
+      })
+    } catch {
+      Log.error(#file, "SQLite Error: Could not create table")
+      return
+    }
+    
+    // Update (not insert) if uniqueID and libraryID match existing row in table
+    let query = sqlTable.filter(sqlLibraryID == libraryID && sqlUpdateID == updateID)
+                        .filter(sqlUpdateID != nil)
+    
+    do {
+      // Update row
+      if try db.run(query.update(sqlParameters <- parameters, sqlHeader <- headerData)) > 0 {
+        Log.info(#file, "SQLite Row Updated - Success")
+        
+      // Insert new row
+      } else {
+        do {
+          try db.run(sqlTable.insert(sqlLibraryID <- libraryID, sqlUpdateID <- updateID, sqlUrl <- urlString, sqlMethod <- methodString, sqlParameters <- parameters, sqlHeader <- headerData, sqlRetries <- 0))
+          Log.info(#file, "SQLite Row Added - Success")
+        } catch {
+          Log.error(#file, "SQLite Error: Could not update table")
+        }
+      }
+    } catch {
+      Log.error(#file, "SQLite Error: Could not update queue")
+    }
+  }
+  
+  class func retryQueue()
+  {
+    guard let db = startDatabaseConnection() else { return }
+    
+    do {
+      for row in try db.prepare(sqlTable) {
+        self.retry(db: db, requestRow: row)
+      }
+    } catch {
+      Log.error(#file, "SQLite Error accessing table")
+    }
+  }
+  
+  
+  // MARK: - Private Functions
+  
+  private class func retry(db: Connection, requestRow: Row)
+  {
+    if (Int(requestRow[sqlRetries]) > QueueMaxRetryCount) {
+      deleteRow(db: db, id: Int(requestRow[sqlID]))
+      Log.info(#file, "Removing after \(Int(requestRow[sqlRetries])) retries")
+      return
+    }
+    
+    do {
+      let ID = Int(requestRow[sqlID])
+      let newValue = Int(requestRow[sqlRetries]) + 1
```

Comment:
It seems like it'll be really easy to max out the allowed retries if the user is offline or has an intermittent connection or if the server goes down for awhile. While I do agree that events that are never going to send should be dropped from the queue eventually, I think another approach is needed here. Perhaps we should only be removing events when the server has returned a particular status code indicating malformation or similar? Maybe that plus a limit in terms of time rather than attempts?